packages/babel-plugin-jest-hoist/src/index.ts

```diff
@@ -152,38 +166,131 @@ FUNCTIONS.deepUnmock = args => args.length === 1 && args[0].isStringLiteral();
-  const shouldHoistExpression = (expr: NodePath): boolean => {
-    if (!expr.isCallExpression()) {
-      return false;
-    }
+const createJestObjectGetter = statement`
+function GETTER_NAME() {
+  const { JEST_GLOBALS_MODULE_JEST_EXPORT_NAME } = require("JEST_GLOBALS_MODULE_NAME");
+  GETTER_NAME = () => JEST_GLOBALS_MODULE_JEST_EXPORT_NAME;
+  return JEST_GLOBALS_MODULE_JEST_EXPORT_NAME;
+}
+`;
 
-    // TODO: avoid type casts - the types can be arrays (is it possible to ignore that without casting?)
-    const callee = expr.get('callee') as NodePath;
-    const expressionArguments = expr.get('arguments');
-    const object = callee.get('object') as NodePath;
-    const property = callee.get('property') as NodePath;
-    return (
-      property.isIdentifier() &&
-      FUNCTIONS[property.node.name] &&
-      (object.isIdentifier(JEST_GLOBAL) ||
-        (callee.isMemberExpression() && shouldHoistExpression(object))) &&
-      FUNCTIONS[property.node.name](
-        Array.isArray(expressionArguments)
-          ? expressionArguments
-          : [expressionArguments],
-      )
-    );
-  };
+const isJestObject = (expression: NodePath<Expression>): boolean => {
+  // global
+  if (
+    expression.isIdentifier() &&
+    expression.node.name === JEST_GLOBAL_NAME &&
+    !expression.scope.hasBinding(JEST_GLOBAL_NAME)
+  ) {
+    return true;
+  }
+  // import { jest } from '@jest/globals'
+  if (
+    expression.referencesImport(
+      JEST_GLOBALS_MODULE_NAME,
+      JEST_GLOBALS_MODULE_JEST_EXPORT_NAME,
+    )
+  ) {
+    return true;
+  }
+  // import * as JestGlobals from '@jest/globals'
+  if (
+    expression.isMemberExpression() &&
+    !expression.node.computed &&
+    expression
+      .get<'object'>('object')
+      .referencesImport(JEST_GLOBALS_MODULE_NAME, '*') &&
+    expression.node.property.name === JEST_GLOBALS_MODULE_JEST_EXPORT_NAME
+  ) {
+    return true;
+  }
 
-  const visitor: Visitor = {
-    ExpressionStatement(path) {
-      if (shouldHoistExpression(path.get('expression') as NodePath)) {
-        // @ts-ignore: private, magical property
-        path.node._blockHoist = Infinity;
-      }
-    },
-  };
+  return false;
+};
 
-  return {visitor};
+const extractJestObjExprIfHoistable = <T extends Node>(
+  expr: NodePath<T>,
+): NodePath<Expression> | null => {
+  if (!expr.isCallExpression()) {
+    return null;
+  }
+
+  const callee = expr.get<'callee'>('callee');
+  const args = expr.get<'arguments'>('arguments');
+
+  if (!callee.isMemberExpression()) {
+    return null;
+  }
+
+  const object = callee.get<'object'>('object');
+  const property = callee.get<'property'>('property') as
+    | NodePath<Expression>
+    | NodePath<Identifier>;
+
+  if (!property.isIdentifier()) {
+    return null;
+  }
+  const propertyName = property.node.name;
+
+  const jestObjExpr = isJestObject(object)
+    ? object
+    : // The Jest object could be returned from another call since the functions are all chainable.
```

Comment:
do we have tests for this? `jest.useFakeTimers().mock('hahah!')`?

Reply:
For the mock functions among each other yes - non-mocking-related functions have never worked. But we wouldn't wanna hoist other calls, right? Could change the meaning

Reply:
very true